examples/example-wave.cpp

```diff
@@ -0,0 +1,271 @@
+//   this list of conditions and the disclaimer below.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the disclaimer (as noted below) in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of the LLNS/LLNL nor the names of its contributors may
+//   be used to endorse or promote products derived from this software without
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include <cmath>
+#include <cstdlib>
+#include <iostream>
+#include "memoryManager.hpp"
+
+#include "RAJA/RAJA.hpp"
+#include "RAJA/util/defines.hpp"
+
+const int sr    = 2;  // Stencil radius
+const double PI = 3.14159265359;
+
+struct grid_s {
+  double ox, dx;
+  int nx;
+};
+
+
+/*
+  Wave propagator
+*/
+template <typename T, typename fdNestedPolicy>
+void Wave(T *P1, T *P2, RAJA::RangeSegment fdBounds, double ct, int nx)
+{
+  RAJA::forallN<fdNestedPolicy>(
+    fdBounds, fdBounds, [=] RAJA_HOST_DEVICE(RAJA::Index_type ty, RAJA::Index_type tx) {  
+      
+      /*
+        Coefficients for a fourth order stencil
+      */
+      double coeff[5] = {
+        -1.0 / 12.0, 4.0 / 3.0, -5.0 / 2.0, 4.0 / 3.0, -1.0 / 12.0};
+      
+      const int id = tx + ty * nx;
+      double P_old = P1[id];
+      double P_curr = P2[id];
+
+      /*
+        Compute Laplacian
+       */
+      double lap = 0.0;
+      
+      for (auto r : RAJA::RangeSegment(-sr, sr + 1)) {
+        const int xi = (tx + r + nx) % nx;
+        const int idx = xi + nx * ty;
+        lap += coeff[r + sr] * P2[idx];
+        
+        const int yi = (ty + r + nx) % nx;
+        const int idy = tx + nx * yi;
+        lap += coeff[r + sr] * P2[idy];
+      }
+      
+      P1[id] = 2 * P_curr - P_old + ct * lap;
+      
+    });
+}
+
+double waveSol(double t, double x, double y);
+void setIC(double *P1, double *P2, double t0, double t1, grid_s grid);
+void computeErr(double *P, double tf, grid_s grid);
+
+/*
```

Comment:
I think it's better to explain the example up top before any substantial code.  Also, explain the RAJA features used in the example here as well.